### PR TITLE
Add in-place RPATH replacement

### PR DESCRIPTION
**Pull request by haampie**
Whenever the rpath string actually _grows_, it falls back to patchelf,
when it stays the same length or gets shorter, we update it in-place,
padded with null bytes.

This PR only deals with absolute -> absolute rpath replacement. We don't
use `_build_tarball(relative=True)` in our CI. If `relative` then it falls
back to the old replacement code.

With this PR, relocation time goes down significantly, likely because patchelf
does some odd things with mmap, causing lots of overhead. Example:

- `binutils`: 700MB installed, goes from `1.91s` to `0.57s`, or `3.4x` faster.
   Relocation time: 27% -> 10% of total install time
- `llvm`: 6.8GB installed, goes from `28.56s` to `5.38`, or `5.3x` faster.
   Relocation time: 44% -> 13% of total install time

The bottleneck is now decompression.

### CODE REVIEW

**Comment by haampie** [comment]
With this:

```python
def _elf_rpaths_for(path):
    return spack.object_file.elf.get_rpaths(path)


def _set_elf_rpaths(target, rpaths):
    assert spack.object_file.elf.replace_rpaths(target, rpaths)
```

```
$ spack clean -db && time spack spec zlib
==> Removing bootstrapped software and configuration in "/tmp/tmp.ylfs25IL9g/bootstrap"
Input spec
--------------------------------
zlib

Concretized
--------------------------------
==> Bootstrapping clingo from pre-built binaries
zlib@1.2.11%gcc@10.3.0+optimize+pic+shared arch=linux-ubuntu20.04-zen2


real	0m2.405s
```

:upside_down_face:

**Comment by haampie** [comment]
https://github.com/spack/spack/pull/27616 is required for this, so rebased on top of that. (now merged)

**Comment by haampie** [comment]
On a slower filesystem:
```console

# before
$ ./spack/bin/spack clean -db
$ time ./spack/bin/spack spec zlib
real	0m30.651s
user	0m13.970s
sys	0m2.632s


# after
$ ./spack/bin/spack clean -db
$ time ./spack/bin/spack spec zlib
real	0m14.648s
user	0m5.419s
sys	0m1.395s
```

(I guess this also includes the time of `spack compiler find`, but I haven't really checked that yet)

**Review by bernhardkaindl — DISMISSED** [review]
My approval - But I only reviewed the in-place update of the rpath string in ELF.

LGTM! I worked on ELF files/parsing/modifying before, I can attest that this works and is safe.

The other aspects are better reviewed by a spack core developer of course!

**Comment by haampie** [comment]
Some defensive programming (`f.read(n)` actually returning `n` bytes, and certain offsets/sizes not being `0`) in various places wouldn't hurt, I'll add that later. Edit: done.

**Comment by haampie** [comment]
I've added a --force-rpath equivalent now, as well as making the python-version exit when multiple DT_RPATH/DT_RUNPATH tags exist in the dynamic array, so that we fall back to patchelf in those cases.

**Comment by haampie** [comment]
Installing LLVM from a binary cache is about 10 seconds faster, note that the bulk of time (50s) is spent on looking for prefix paths in the binary, so it doesn't feel a lot faster. (fwiw, multithreaded zstd, find and replace of strings disabled, and this PR makes LLVM install in 4 seconds from the binary cache).

I also noticed that `patchelf` can't handle some of the cases that I thought it could:

```
==> [2021-11-26-17:03:36.248942] '/home/user/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-10.3.0/patchelf-0.13-djm2fh7ipgk2vm6ndxqqof4uft2rroai/bin/patchelf' '--force-rpath' '--set-rpath' '' '/tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libarcher.so'
==> [2021-11-26-17:03:36.365471] Warning: patchelf --force-rpath --set-rpath /tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libarcher.so failed with error Command exited with status -11:
    '/home/user/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-10.3.0/patchelf-0.13-djm2fh7ipgk2vm6ndxqqof4uft2rroai/bin/patchelf' '--force-rpath' '--set-rpath' '' '/tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libarcher.so'
==> [2021-11-26-17:03:36.371928] '/home/user/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-10.3.0/patchelf-0.13-djm2fh7ipgk2vm6ndxqqof4uft2rroai/bin/patchelf' '--force-rpath' '--set-rpath' '' '/tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libomp.so'
==> [2021-11-26-17:03:36.491700] Warning: patchelf --force-rpath --set-rpath /tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libomp.so failed with error Command exited with status -11:
    '/home/user/.spack/bootstrap/store/linux-ubuntu20.04-x86_64/gcc-10.3.0/patchelf-0.13-djm2fh7ipgk2vm6ndxqqof4uft2rroai/bin/patchelf' '--force-rpath' '--set-rpath' '' '/tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libomp.so'
```

Actually, it turns out we're setting an empty string for the rpath, that seems... redundant. In fact:

```
$ patchelf --force-rpath --set-rpath '' /tmp/tmp.82RfP3XQ0J/linux-ubuntu20.04-zen2/gcc-10.3.0/llvm-12.0.1-7cwfjliqenxix7axw3fxxr65wda6p34m/lib/libarcher.so
Segmentation fault (core dumped)
```

:D great! That bug doesn't exist in the Python version:

```
harmen-desktop tmp.dy5NJ9b5gZ $ echo 'int _start(){return 0;}' | gcc -Wl,-rpath,/test -nostdlib -x c -
harmen-desktop tmp.dy5NJ9b5gZ $ readelf -d a.out | grep RUNPA
 0x000000000000001d (RUNPATH)            Library runpath: [/test]
harmen-desktop tmp.dy5NJ9b5gZ $ spack python -c 'spack.object_file.elf.replace_rpaths("a.out", [])'
True
harmen-desktop tmp.dy5NJ9b5gZ $ readelf -d a.out | grep RUNPA
 0x000000000000001d (RUNPATH)            Library runpath: []
```

File: lib/spack/spack/object_file/elf.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+    virtual_addresses = []
+
+    found = False
+    for _ in range(e_phnum):
+        prog_header = f.read(e_phentsize)
+        if len(prog_header) != e_phentsize:
+            raise ElfParsingError("Malformed program header")
+        p_type, = unpack_from(t_word, buffer=prog_header, offset=0)
+
+        # Find PT_LOAD to get the
```
Comment (bernhardkaindl) [inline lib/spack/spack/object_file/elf.py]: @haampie - Unless my guess is wrong our you know a better way to comment this:
```suggestion
        # For PT_LOAD entries: Save the offsets and virtual addrs of the loaded ELF segments
```

**Review by bernhardkaindl — CHANGES_REQUESTED** [review]

**Review by bernhardkaindl — COMMENTED** [review]
Hi Harmen, I applied my guess to complete the comment so we can merge this soon.

**Review by bernhardkaindl — DISMISSED** [review]

File: lib/spack/spack/object_file/elf.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+    virtual_addresses = []
+
+    found = False
+    for _ in range(e_phnum):
+        prog_header = f.read(e_phentsize)
+        if len(prog_header) != e_phentsize:
+            raise ElfParsingError("Malformed program header")
+        p_type, = unpack_from(t_word, buffer=prog_header, offset=0)
+
+        # For PT_LOAD entries: Save the offsets and virtual addrs of the loaded ELF segments
```
Comment (bernhardkaindl) [inline lib/spack/spack/object_file/elf.py]:
```suggestion
        # For PT_LOAD entries: Save offsets and virtual addrs of the loaded ELF segments
```

**Comment by haampie** [comment]
Moved the ELF parsing business into #33628, then this can be a separate PR that just does pure-python rpath replacement.

**Comment by haampie** [comment]
PR now rebased on develop

**Comment by haampie** [comment]
@spackbot rebuild everything

**Comment by haampie** [comment]
@scottwittenburg gitlab CI ran on the wrong commit again. This is a huge problem...

**Comment by haampie** [comment]
@spackbot rebuild everything

**Comment by haampie** [comment]
@spackbot rebuild everything

**Comment by haampie** [comment]
@tgamblin: this successfully built the e4s pipeline with 1,000+ jobs without the need for `patchelf`, which makes it good to go. Note that linking requires rpaths to be set correctly too, it's not just a runtime thing, so it's thoroughly tested.

The failures are all `silo`, which is due to a the bug on develop fixed by c716c6ca95e300ac247f668237fc13f75e1b5289.

I'm adding one more commit to fall back to the old relocation code in case of `rel=True`.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**